File: library/sync-tasks/src/main/java/org/cru/godtools/sync/task/BaseDataSyncTasks.java
package org.cru.godtools.sync.task;

import android.content.Context;

import org.ccci.gto.android.common.db.Query;
import org.ccci.gto.android.common.jsonapi.util.Includes;
import org.cru.godtools.model.Attachment;
import org.cru.godtools.model.Base;
import org.cru.godtools.model.Language;
import org.cru.godtools.model.Tool;
import org.cru.godtools.model.Translation;
import org.cru.godtools.model.event.AttachmentUpdateEvent;
import org.cru.godtools.model.event.LanguageUpdateEvent;
import org.cru.godtools.model.event.ToolUpdateEvent;
import org.cru.godtools.model.event.TranslationUpdateEvent;
import org.keynote.godtools.android.db.Contract.AttachmentTable;
import org.keynote.godtools.android.db.Contract.LanguageTable;
import org.keynote.godtools.android.db.Contract.ToolTable;
import org.keynote.godtools.android.db.Contract.TranslationTable;

import java.util.List;
import java.util.Locale;
import java.util.Map;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;
import androidx.annotation.VisibleForTesting;
import androidx.collection.LongSparseArray;
import androidx.collection.SimpleArrayMap;

import static android.database.sqlite.SQLiteDatabase.CONFLICT_REPLACE;

@RestrictTo(RestrictTo.Scope.LIBRARY)
public abstract class BaseDataSyncTasks extends BaseSyncTasks {
    private static final String[] API_FIELDS_LANGUAGE = {LanguageTable.COLUMN_ID, LanguageTable.COLUMN_NAME};
    private static final String[] API_FIELDS_TOOL =
            {ToolTable.COLUMN_CODE, ToolTable.COLUMN_TYPE, ToolTable.COLUMN_NAME, ToolTable.COLUMN_DESCRIPTION,
                    ToolTable.COLUMN_SHARES, ToolTable.COLUMN_BANNER, ToolTable.COLUMN_DETAILS_BANNER,
                    ToolTable.COLUMN_COPYRIGHT, ToolTable.COLUMN_OVERVIEW_VIDEO};
    private static final String[] API_FIELDS_ATTACHMENT =
            {AttachmentTable.COLUMN_TOOL, AttachmentTable.COLUMN_FILENAME, AttachmentTable.COLUMN_SHA256};
    private static final String[] API_FIELDS_TRANSLATION =
            {TranslationTable.COLUMN_TOOL, TranslationTable.COLUMN_LANGUAGE, TranslationTable.COLUMN_VERSION,
                    TranslationTable.COLUMN_NAME, TranslationTable.COLUMN_DESCRIPTION, TranslationTable.COLUMN_TAGLINE,
                    TranslationTable.COLUMN_MANIFEST, TranslationTable.COLUMN_PUBLISHED};

    BaseDataSyncTasks(@NonNull Context context) {
        super(context);
    }

    void storeLanguages(@NonNull final SimpleArrayMap<Class<?>, Object> events, @NonNull final List<Language> languages,
                        @Nullable final Map<Locale, Language> existing) {
        for (final Language language : languages) {
            if (existing != null) {
                existing.remove(language.getCode());
            }
            storeLanguage(events, language);
        }

        // prune any existing languages that weren't synced and aren't already added to the device
        if (existing != null) {
            for (final Language language : existing.values()) {
                if (language != null && !language.isAdded()) {
                    mDao.delete(language);
                    coalesceEvent(events, LanguageUpdateEvent.INSTANCE);
                }
            }
        }
    }

    @VisibleForTesting
    void storeLanguage(@NonNull final SimpleArrayMap<Class<?>, Object> events, @NonNull final Language language) {
        // this language doesn't exist yet, check to see if a different language shares the same id
        if (language.getId() != Base.INVALID_ID && mDao.refresh(language) == null) {
            // update the language code to preserve the added state
            mDao.streamCompat(Query.select(Language.class).where(LanguageTable.FIELD_ID.eq(language.getId())).limit(1))
                    .findFirst()
                    .ifPresent(old -> {
                        mDao.update(language, mDao.getPrimaryKeyWhere(old), LanguageTable.COLUMN_CODE);
                        coalesceEvent(events, LanguageUpdateEvent.INSTANCE);
                    });
        }

        mDao.updateOrInsert(language, CONFLICT_REPLACE, API_FIELDS_LANGUAGE);
        coalesceEvent(events, LanguageUpdateEvent.INSTANCE);
    }

    void storeTools(@NonNull final SimpleArrayMap<Class<?>, Object> events, @NonNull final List<Tool> tools,
                    @Nullable final LongSparseArray<Tool> existing, @NonNull final Includes includes) {
        for (final Tool tool : tools) {
            if (existing != null) {
                existing.remove(tool.getId());
            }
            storeTool(events, tool, includes);
        }

        // prune any existing tools that weren't synced and aren't already added to the device
        if (existing != null) {
            for (int i = 0; i < existing.size(); i++) {
                final Tool tool = existing.valueAt(i);
                if (!tool.isAdded()) {
                    mDao.delete(tool);
                    coalesceEvent(events, ToolUpdateEvent.INSTANCE);

                    // delete any attachments for this tool
                    mDao.delete(Attachment.class, AttachmentTable.FIELD_TOOL.eq(tool.getId()));
                }
            }
        }
    }

    private void storeTool(@NonNull final SimpleArrayMap<Class<?>, Object> events, @NonNull final Tool tool,
                           @NonNull final Includes includes) {
        mDao.updateOrInsert(tool, CONFLICT_REPLACE, API_FIELDS_TOOL);
        coalesceEvent(events, ToolUpdateEvent.INSTANCE);

        // persist any related included objects
        if (includes.include(Tool.JSON_LATEST_TRANSLATIONS)) {
            final List<Translation> translations = tool.getLatestTranslations();
            if (translations != null) {
                final LongSparseArray<Translation> existing;
                if (tool.getCode() != null) {
                    existing = index(mDao.get(
                            Query.select(Translation.class).where(TranslationTable.FIELD_TOOL.eq(tool.getCode()))));
                } else {
                    existing = null;
                }
                storeTranslations(events, translations, existing, includes.descendant(Tool.JSON_LATEST_TRANSLATIONS));
            }
        }
        if (includes.include(Tool.JSON_ATTACHMENTS)) {
            final List<Attachment> attachments = tool.getAttachments();
            if (attachments != null) {
                final LongSparseArray<Attachment> existing = index(mDao.get(
                        Query.select(Attachment.class).where(AttachmentTable.FIELD_TOOL.eq(tool.getId()))));
                storeAttachments(events, attachments, existing, includes.descendant(Tool.JSON_ATTACHMENTS));
            }
        }
    }

    private void storeTranslations(@NonNull final SimpleArrayMap<Class<?>, Object> events,
                                   @NonNull final List<Translation> translations,
                                   @Nullable final LongSparseArray<Translation> existing,
                                   @NonNull final Includes includes) {
        for (final Translation translation : translations) {
            if (existing != null) {
                existing.remove(translation.getId());
            }
            storeTranslation(events, translation, includes);
        }

        // prune any existing translations that weren't synced and aren't downloaded to the device
        if (existing != null) {
            for (int i = 0; i < existing.size(); i++) {
                final Translation translation = mDao.refresh(existing.valueAt(i));
                if (translation != null && !translation.isDownloaded()) {
                    mDao.delete(translation);
                    coalesceEvent(events, TranslationUpdateEvent.INSTANCE);
                }
            }
        }
    }

    private void storeTranslation(@NonNull final SimpleArrayMap<Class<?>, Object> events,
                                  @NonNull final Translation translation, @NonNull final Includes includes) {
        mDao.updateOrInsert(translation, API_FIELDS_TRANSLATION);
        coalesceEvent(events, TranslationUpdateEvent.INSTANCE);

        if (includes.include(Translation.JSON_LANGUAGE)) {
            final Language language = translation.getLanguage();
            if (language != null) {
                storeLanguage(events, language);
            }
        }
    }

    private void storeAttachments(@NonNull final SimpleArrayMap<Class<?>, Object> events,
                                  @NonNull final List<Attachment> attachments,
                                  @Nullable final LongSparseArray<Attachment> existing,
                                  @NonNull final Includes includes) {
        for (final Attachment attachment : attachments) {
            if (existing != null) {
                existing.remove(attachment.getId());
            }
            storeAttachment(events, attachment, includes);
        }

        // prune any existing attachments that weren't synced
        if (existing != null) {
            for (int i = 0; i < existing.size(); i++) {
                final Attachment attachment = existing.valueAt(i);
                mDao.delete(attachment);
                coalesceEvent(events, AttachmentUpdateEvent.INSTANCE);
            }
        }
    }

    private void storeAttachment(@NonNull final SimpleArrayMap<Class<?>, Object> events,
                                 @NonNull final Attachment attachment, @NonNull final Includes includes) {
        mDao.updateOrInsert(attachment, API_FIELDS_ATTACHMENT);
        coalesceEvent(events, AttachmentUpdateEvent.INSTANCE);
    }
}

File: library/sync-tasks/src/main/java/org/cru/godtools/sync/task/BaseDataSyncTasks.kt
package org.cru.godtools.sync.task

import android.database.sqlite.SQLiteDatabase
import androidx.annotation.RestrictTo
import androidx.annotation.VisibleForTesting
import androidx.collection.LongSparseArray
import androidx.collection.SimpleArrayMap
import androidx.collection.forEach
import org.ccci.gto.android.common.db.Query
import org.ccci.gto.android.common.db.get
import org.ccci.gto.android.common.jsonapi.util.Includes
import org.cru.godtools.model.Attachment
import org.cru.godtools.model.Base
import org.cru.godtools.model.Language
import org.cru.godtools.model.Tool
import org.cru.godtools.model.Translation
import org.cru.godtools.model.event.AttachmentUpdateEvent
import org.cru.godtools.model.event.LanguageUpdateEvent
import org.cru.godtools.model.event.ToolUpdateEvent
import org.cru.godtools.model.event.TranslationUpdateEvent
import org.greenrobot.eventbus.EventBus
import org.keynote.godtools.android.db.Contract.AttachmentTable
import org.keynote.godtools.android.db.Contract.LanguageTable
import org.keynote.godtools.android.db.Contract.ToolTable
import org.keynote.godtools.android.db.Contract.TranslationTable
import org.keynote.godtools.android.db.GodToolsDao
import java.util.Locale

@RestrictTo(RestrictTo.Scope.LIBRARY)
abstract class BaseDataSyncTasks internal constructor(protected val dao: GodToolsDao, eventBus: EventBus) :
    BaseSyncTasks(eventBus) {
    // region Tools
    protected fun storeTools(
        events: SimpleArrayMap<Class<*>, Any>,
        tools: List<Tool>,
        existing: LongSparseArray<Tool>?,
        includes: Includes
    ) {
        tools.forEach {
            storeTool(events, it, includes)
            existing?.remove(it.id)
        }

        // prune any existing tools that weren't synced and aren't already added to the device
        existing?.forEach { _, tool ->
            if (tool.isAdded) return@forEach

            dao.delete(tool)
            coalesceEvent(events, ToolUpdateEvent)

            // delete any attachments for this tool
            dao.delete(Attachment::class.java, AttachmentTable.FIELD_TOOL.eq(tool.id))
        }
    }

    private fun storeTool(events: SimpleArrayMap<Class<*>, Any>, tool: Tool, includes: Includes) {
        dao.updateOrInsert(
            tool, SQLiteDatabase.CONFLICT_REPLACE,
            ToolTable.COLUMN_CODE, ToolTable.COLUMN_TYPE, ToolTable.COLUMN_NAME, ToolTable.COLUMN_DESCRIPTION,
            ToolTable.COLUMN_SHARES, ToolTable.COLUMN_BANNER, ToolTable.COLUMN_DETAILS_BANNER,
            ToolTable.COLUMN_DEFAULT_ORDER, ToolTable.COLUMN_OVERVIEW_VIDEO
        )
        coalesceEvent(events, ToolUpdateEvent)

        // persist related included objects
        if (includes.include(Tool.JSON_LATEST_TRANSLATIONS)) tool.latestTranslations?.let { translations ->
            storeTranslations(
                events, translations,
                includes = includes.descendant(Tool.JSON_LATEST_TRANSLATIONS),
                existing = tool.code?.let { code ->
                    index(Query.select<Translation>().where(TranslationTable.FIELD_TOOL.eq(code)).get(dao))
                }
            )
        }
        if (includes.include(Tool.JSON_ATTACHMENTS)) tool.attachments?.let { attachments ->
            storeAttachments(
                events, attachments,
                existing = index(Query.select<Attachment>().where(AttachmentTable.FIELD_TOOL.eq(tool.id)).get(dao))
            )
        }
    }
    // endregion Tools

    // region Languages
    protected fun storeLanguages(
        events: SimpleArrayMap<Class<*>, Any>,
        languages: List<Language>,
        existing: MutableMap<Locale, Language>?
    ) {
        languages.forEach {
            storeLanguage(events, it)
            existing?.remove(it.code)
        }

        // prune any existing languages that weren't synced and aren't already added to the device
        existing?.values?.forEach {
            dao.delete(it)
            coalesceEvent(events, LanguageUpdateEvent)
        }
    }

    @VisibleForTesting
    protected fun storeLanguage(events: SimpleArrayMap<Class<*>, Any>, language: Language) {
        // this language doesn't exist yet, check to see if a different language shares the same id
        if (language.id != Base.INVALID_ID && dao.refresh(language) == null) {
            // update the language code to preserve the added state
            Query.select<Language>().where(LanguageTable.FIELD_ID.eq(language.id)).limit(1)
                .get(dao)
                .firstOrNull()
                ?.let { old ->
                    dao.update(language, dao.getPrimaryKeyWhere(old), LanguageTable.COLUMN_CODE)
                    coalesceEvent(events, LanguageUpdateEvent)
                }
        }

        dao.updateOrInsert(
            language, SQLiteDatabase.CONFLICT_REPLACE,
            LanguageTable.COLUMN_ID, LanguageTable.COLUMN_NAME
        )
        coalesceEvent(events, LanguageUpdateEvent)
    }
    // endregion Languages

    // region Translations
    private fun storeTranslations(
        events: SimpleArrayMap<Class<*>, Any>,
        translations: List<Translation>,
        existing: LongSparseArray<Translation>?,
        includes: Includes
    ) {
        translations.forEach {
            storeTranslation(events, it, includes)
            existing?.remove(it.id)
        }

        // prune any existing translations that weren't synced and aren't downloaded to the device
        existing?.forEach { _, translation ->
            dao.refresh(translation)?.takeUnless { it.isDownloaded }?.let {
                dao.delete(it)
                coalesceEvent(events, TranslationUpdateEvent)
            }
        }
    }

    private fun storeTranslation(events: SimpleArrayMap<Class<*>, Any>, translation: Translation, includes: Includes) {
        dao.updateOrInsert(
            translation,
            TranslationTable.COLUMN_TOOL, TranslationTable.COLUMN_LANGUAGE, TranslationTable.COLUMN_VERSION,
            TranslationTable.COLUMN_NAME, TranslationTable.COLUMN_DESCRIPTION, TranslationTable.COLUMN_TAGLINE,
            TranslationTable.COLUMN_MANIFEST, TranslationTable.COLUMN_PUBLISHED
        )
        coalesceEvent(events, TranslationUpdateEvent)

        if (includes.include(Translation.JSON_LANGUAGE)) translation.language?.let { storeLanguage(events, it) }
    }
    // endregion Translations

    // region Attachments
    private fun storeAttachments(
        events: SimpleArrayMap<Class<*>, Any>,
        attachments: List<Attachment>,
        existing: LongSparseArray<Attachment>?
    ) {
        attachments.forEach {
            storeAttachment(events, it)
            existing?.remove(it.id)
        }

        // prune any existing attachments that weren't synced
        existing?.forEach { _, attachment ->
            dao.delete(attachment)
            coalesceEvent(events, AttachmentUpdateEvent)
        }
    }

    private fun storeAttachment(events: SimpleArrayMap<Class<*>, Any>, attachment: Attachment) {
        dao.updateOrInsert(
            attachment,
            AttachmentTable.COLUMN_TOOL, AttachmentTable.COLUMN_FILENAME, AttachmentTable.COLUMN_SHA256
        )
        coalesceEvent(events, AttachmentUpdateEvent)
    }
    // endregion Attachments
}
